src/cascade/model/model_reader.py

```diff
@@ -0,0 +1,313 @@
+
+    Args:
+        dismod_file: The DismodFile wrapper.
+
+    Returns:
+        DataFrame: Columns are ``name``, ``avg_integrand``,
+        ``weighted_residual``.
+    """
+    # The data table associates data_id with data_name, but data_subset
+    # has an id for all data that were used and will have residuals.
+    data_subset = dismod_file.data_subset.reset_index(drop=True)
+    subset_id_to_name = data_subset.merge(dismod_file.data[["data_id", "data_name"]], on="data_id") \
+        .drop(columns=["data_id"])
+    return dismod_file.fit_data_subset.reset_index(drop=True).merge(
+        subset_id_to_name, left_on="fit_data_subset_id", right_on="data_subset_id") \
+        .drop(columns=["data_subset_id", "fit_data_subset_id"]) \
+        .rename(columns={"data_name": "name"})
+
+
+def read_var_table_as_id(dismod_file):
+    """
+    This reads the var table in order to find the ids for all of the vars.
+    It puts those into a DismodGroups which can then decode any table
+    associated with the var table.
+    The empty vars come from the Model.model_variables property.
+    This ``var_id`` table has ``age`` and ``time`` and uses real
+    locations instead of ``node_id``. It's meant for joins with tables
+    in the Dismod file.
+    """
+    parent_node = read_parent_node(dismod_file)
+    child_node = read_child_nodes(dismod_file, read_parent_node(dismod_file))
+
+    # The var table has a random field for each smooth_id, except for the
+    # random effects. It doesn't identify which random field each smooth id
+    # goes with, so we first make the DismodGroups structure of smooths
+    # and then invert it in order to read the vars table.
+    inverted_smooth = read_inverted_smooths(dismod_file, parent_node, child_node)
+
+    # This groupby to separate the smooth grids is split into two levels because
+    # the second level is node_id, which is nan for mulcovs, and groupby
+    # excludes rows with nans in the keys.
+    var_ids = DismodGroups()
+    age, time = (dismod_file.age, dismod_file.time)
+    for smooth_id, sub_grid_df in dismod_file.var.groupby(["smooth_id"]):
+        if sub_grid_df[sub_grid_df.var_type == "rate"].empty:
+            group_name, key = inverted_smooth[(smooth_id, parent_node)]
+            var_ids[group_name][key] = _add_one_field_to_vars(sub_grid_df, age, time)
+        else:
+            # Multiple random effects, identified as "rates," can share a smooth.
+            for node_id, re_grid_df in sub_grid_df.groupby(["node_id"]):
+                group_name, key = inverted_smooth[(smooth_id, node_id)]
+                var_ids[group_name][key] = _add_one_field_to_vars(re_grid_df, age, time)
+
+    if var_ids.count() != len(dismod_file.var):
+        MATHLOG.error(f"Found {var_ids.count()} of {len(dismod_file.var)} vars in db file.")
+
+    return rename_node_to_location(dismod_file.node, var_ids)
+
+
+def _add_one_field_to_vars(sub_grid_df, age, time):
```

Comment:
Similar to a comment above. This extracts data from a data frame, it doesn't add anything so the name is confusing.